packages/flutter_tools/lib/src/runner/flutter_command_runner.dart

```diff
@@ -390,4 +390,33 @@ class FlutterCommandRunner extends CommandRunner {
     if (ArtifactStore.flutterRoot == null)
       ArtifactStore.flutterRoot = _defaultFlutterRoot;
   }
+
+  /// Get all pub packages in the Flutter repo.
+  List<Directory> getRepoPackages() {
+    String root = path.absolute(ArtifactStore.flutterRoot);
+    List<String> directories = <String>[];
+
+    // dev/ (excluding remove dev/docs)
+    directories.addAll(_gatherProjectPaths(path.join(root, 'dev')));
+    directories.removeWhere((String dir) => dir.endsWith('docs'));
+
+    // examples/
+    directories.addAll(_gatherProjectPaths(path.join(root, 'examples')));
+
+    // packages/
+    directories.addAll(_gatherProjectPaths(path.join(root, 'packages')));
+
+    return directories.map((String dir) => new Directory(dir)).toList();
+  }
+
+  static List<String> _gatherProjectPaths(String rootPath) {
+    if (FileSystemEntity.isFileSync(path.join(rootPath, 'pubspec.yaml')))
+      return <String>[rootPath];
+
+    return new Directory(rootPath)
+      .listSync(followLinks: false)
+      .expand((FileSystemEntity entity) {
+        return entity is Directory ? _gatherProjectPaths(entity.path) : <String>[];
+      });
+  }
```

Comment:
Yeah, scanning the whole repo will probably prevent future problems. `.analysis_options` files let you ignore directories, but those files are not read if they don't live next to a pubspec. We could use `.gitignore` files. That might have issues where you don't want directories analyzed, but you do want them committed to git. I've run into that w/ known bad libs (used for testing). Plus any file opening and glob matching may be more work than I want to do during the scan.

Hard-coded exclusions, or marker files as you suggest, might be the simplest solutions here. I'll take another look at this.